### ISSUE TITLE
Insecure require() should set insecure env on file

### ISSUE THREAD

**Issue by rubenwardy** [minetest/minetest#10877]
## Problem

Currently, mods need to use the following work around to get `ie.require` to work:

```lua
local tmp = require
require = ie.require 
require("module")
require = tmp
```

This isn't secure, because a malicious mod could set a `__newindex` on _G and capture require assignments.

## Solutions

Minetest should provide its own require for the insecure environment which sets the insecure env on the file

## Alternatives

TBD

**Comment by Desour** [comment]
There are also other reasons why this is insecure, ie. the requested module sets globals in `_G`, or uses overwritten functions from `_G`.

According to <https://www.lua.org/manual/5.1/manual.html#2.9>, what one wants to do is probably to set the thread (=global) environment:
>Threads are created sharing the environment of the creating thread. Userdata and C functions are created sharing the environment of the creating C function. Non-nested Lua functions (created by [loadfile](http://www.lua.org/manual/5.1/manual.html#pdf-loadfile), [loadstring](http://www.lua.org/manual/5.1/manual.html#pdf-loadstring) or [load](http://www.lua.org/manual/5.1/manual.html#pdf-load)) are created sharing the environment of the creating thread. Nested Lua functions are created sharing the environment of the creating Lua function.

Also, the string metatable might be used by the loaded file, and there could be a debug hook.

I came up with the following wrapper for the require function. But I'm not totally sure if it's secure:

```lua
local IE = minetest.request_insecure_environment()
_G.assert(IE ~= nil)

-- the same as `IE.require(...)`, but sets the env to IE
local function internal.require_with_IE_env(...)
	-- be sure that there is no hook, otherwise one could get IE via getfenv
	IE.debug.sethook()

	local old_thread_env = IE.getfenv(0)
	local old_string_metatable = IE.debug.getmetatable("")

	-- set env of thread
	-- (the loader used by IE.require will probably use the thread env for
	-- the loaded functions)
	IE.setfenv(0, IE)

	-- also set the string metatable because the lib might use it while loading
	-- (actually, we probably have to do this every time we call a `require()`d
	-- function, but for performance reasons we only do it if the function
	-- uses the string metatable)
	-- (Maybe it would make sense to set the string metatable __index field
	-- to a function that grabs the string table from the thread env.)
	IE.debug.setmetatable("", {__index = IE.string})

	-- (IE.require's env is neither _G, nor IE. we need to leave it like this,
	-- otherwise it won't find the loaders (it uses the global `loaders`, not
	-- `package.loaders` btw. (see luajit/src/lib_package.c)))

	-- we might be pcall()ed, so we need to pcall to make sure that we reset
	-- the thread env afterwards
	local ok, ret = IE.pcall(IE.require, ...)

	-- reset env of thread
	IE.setfenv(0, old_thread_env)

	-- also reset the string metatable
	IE.debug.setmetatable("", old_string_metatable)

	if not ok then
		IE.error(ret)
	end
	return ret
end
```

Edit: Apparently I'm using this [in dslib](https://codeberg.org/Desour/dslib/src/commit/d92c72856129fea1895be7c16b1e14342b739e4d/init.lua#L118). If I find bugs I'll probably forget to edit this comment.

**Comment by doxygen-spammer** [comment]
I think that I now understand the main problem.
Unfortunately, neither rubenwardy’s nor Desour’s presented workarounds touch this problem.

## Explanation
<details><summary>expand...</summary>
It all comes down to Minetest using the wrong environments.
The “standard” environment for untrusted mods shall be a “sandbox”, but actually Minetest replaces the “global environment” with it.
The “insecure environment” for trusted mods is a non-global environment, and thus appears like an anti-sandbox in a sandbox.

The problem is that there are several ways to access the global environment, and sooner or later stuff will leak into the “sandbox”.

1. require() caches packages, because any package may request any other package as often as needed. And functions/threads running in other environments are still allowed to request packages (unless the environment is a sandbox without require()). Therefore, require() loads all packages in the “global environment”.
    * By the way: require() does not have an environment, because it is a C function. The environment only resolves accesses to global variables. package.loaders is a table, and therefore doesn’t have an environment either. It is just accessible from the main environment. (`loaders` is not a global, its a field in the table at LUA_ENVIRONINDEX. It is accessible through package.loaders and may be written whenever needed. `package`, in turn, is a global in the global environment.)
2. Even if require() loads packages in specific environments, some C modules will still export their (potentially harmful) functionality in the “global environment”—just because they do not know other environments. LuaFileSystem is an example.
3. dofile() runs everything in the “global environment”, no matter from which environment it is called.

There may be confusion about the term “global environment”:

> **load (func [, chunkname])**
> [...] returns the compiled chunk as a function. [...] The environment of the returned function is the global environment.

If you read this, you may think the global environment is that what provides the global variables to my code. This is just the “environment”. The “global environment” is what Lua stores at LUA_RIDX_GLOBALS. (And that is where Minetest puts the “sandbox” for untrusted mods.)
</details>

TL;DR: After calling require() or dofile(), your mod’s code is back in the main environment and connects both environments without your intent.

## Suggestion
Minetest should swap the environments. (Untrusted mods are not affected, they can not even detect this.)

The environment at LUA_RIDX_GLOBALS would be the “insecure environment”, where any module may dump its dangerous functionality.
The sandbox with the `minetest` API, where all mods start execution, would be stored at CUSTOM_RIDX_GLOBALS_UNTRUSTED.
The sandbox must not contain functions that can give access to the global environment. 
(The mesecons_luacontroller mod is a nice example for this.)
Functions like dofile() need to be reimplemented to run code in the sandbox.
(They are actually already reimplemented to limit filesystem access.)

## My workaround
I just made a mod that runs a somewhat complex application in the insecure environment.
It was necessary to reimplement require() and dofile() to run code in the “anti-sandbox”.
It is actually not very complicated, but it took me a while to put everything together.

```lua
local env = minetest.request_insecure_environment();

-- Call this before using anything inside the insecure environment.
local function seal_environment()
    -- debug.sethook() and debug.getinfo() allow to collect
    -- all called functions.
    -- Prevent collecting functions from the insecure environment.
    env.debug.sethook();

    -- As of Minetest 5.7, string is the only type with metatable.
    -- Other mods could modify this metatable to record and modify
    -- string values used inside the insecure environment.
    -- Access to this metatable can not be restricted.
    -- Before entering the insecure environment, the metatable must
    -- be reset. Minetest specific methods (split, trim) can stay.
    local m = env.getmetatable("");
    env.setmetatable(m, nil); -- Remove any spies.
    local mi = m.__index;
    env.setmetatable(mi, nil); -- Remove any spies here too.
    for k, v in env.pairs(env.string) do
        -- Reset standard string methods.
        mi[k] = v;
    end
    for k in env.pairs(m) do
        -- Remove anything else.
        m[k] = nil;
    end
    m.__index = mi;
end

seal_environment();

-- Minetest engine bug #10877:
-- Minetest attemts to make the insecure environment a nonstandard
-- environment. This doesn’t work. C functions, like require(),
-- still work in the standard environment. Modules require()-ing
-- other modules therefore receive the disabled require() of
-- the Minetest environment.
--
-- It is necessary to modify require().

-- This adds a custom loader to require().
-- It will explicitly search for the modules and load them
-- using loadfile().
-- This allows to explicitly set the environment of these chunks,
-- before they are executed for the first time.
--
-- Note that this is only enough to cover Lua modules,
-- but palette_dev has only Lua modules.
local function loader(module_name)
    for path in env.string.gmatch(env.package.path, "[^;]+") do
        local n = env.string.gsub(module_name, "%.", "/");
        local p = env.string.gsub(path, "%?", n);

        if env.setenv then
            -- Lua 5.1
            local l = env.loadfile(p);
            if l then
                return env.setfenv(l, env);
            end
        else
            -- Lua 5.2+
            local l = env.loadfile(p, --[[ mode ]] nil, env);
            if l then
                return l;
            end
        end
    end
end

env.table.insert(env.package.loaders, 1, loader);

-- dofile() runs the loaded file in the standard environment,
-- just like require().
-- Overwrite it to set the environment of the loaded file.
function env.dofile(p)
    if env.setenv then
        -- Lua 5.1
        local l = assert(env.loadfile(p));
        return env.setfenv(l, env)();
    else
        -- Lua 5.2+
        local l = assert(env.loadfile(p, --[[ mode ]] nil, env));
        return l();
    end
end

-- Preload lfs (LuaFileSytem).
-- lfs exports its functions in the standard global environment.
-- Like observed above, this is the Minetest “sandbox” environment.
-- The lfs table needs to be cleared explicitly, so other mods
-- can not do harm with it.
env.rawset(_G, "lfs", 1234); -- Disable __newindex spies.
env.require("lfs");
env.assert(env.rawget(_G, "lfs") ~= 1234);
env.assert(env.lfs == nil);
env.rawset(_G, "lfs", nil);

-- Determine the exact location of this init.lua file.
local init_source = env.debug.getinfo(1, "S").source;
local i_am_here = env.string.sub(init_source, 2, -9);
env.assert(init_source == "@" .. i_am_here .. "init.lua");
env.assert(env.string.sub(i_am_here, 1, 1) == "/");

-- This is now the command handler using the insecure environment.
local command_handler_inner = env.dofile(i_am_here .. "command_handler.lua");

-- The environment must not become accessible through getfenv(),
-- therefore the command handler must be wrapped once.
-- Additionally, seal_environment() is necessary before every entry.
command_handler = function(a, b)
    seal_environment();

    local c, d = command_handler_inner(a, b);
    return c, d;
end
```

require() will still not work to load modules of the application.
command_handler.lua uses the same algorithm to get i_am_here, and adds it to package.path.
From then on, everything runs smooth.

**Comment by Desour** [comment]
> 1. [...] And functions/threads running in other environments are still allowed to request packages (unless the environment is a sandbox without require()).

The normal server modding env doesn't contain `require`.

>2. Even if require() loads packages in specific environments, some C modules will still export their (potentially harmful) functionality in the “global environment”—just because they do not know other environments.

That's why I set the global environment in my approach while the require is done (`IE.setfenv(0, IE)`).
(Btw. why are you using quotes on "global environment"? It's not some vague term, see http://www.lua.org/manual/5.1/manual.html#2.9>.)

>3. dofile() runs everything in the “global environment”, no matter from which environment it is called.

Again, see <http://www.lua.org/manual/5.1/manual.html#2.9>.
If you set the environment of a function, it's just used by that function (and functions created in that function).


Sorry, but it looks to me like you have some misconceptions / lack of knowledge about lua 5.1 environments. I can't see any *new* insight in your comment.
